052-k8s-server-side-apply.md

```diff
@@ -0,0 +1,147 @@
+| x                      | y                      | Strimzi     | yes              | no                  |
+| x                      | not set                | Strimzi     | yes              | no                  |
+| x                      | not set                | Not Strimzi | no               | yes                 |
+| not set                | y                      | Strimzi     | yes              | no                  |
+| not set                | y                      | Not Strimzi | yes              | no                  |
+
+The key thing to take from this table is that currently Strimzi is trying to make changes to ALL fields (minus short `ignorableFields` list), once Strimzi has knowledge of the fields it owns - we don't want it to try and change the fields it doesn't own.
+
+It would be wise to try and include this in as many operators as possible, but in some situations where it is decided against it can be overridden in the code and left as is.
+
+---
+### Modifications to existing behaviours
+
+The only change to existing (non-Strimzi specific) behaviours is that `ResourceDiff` won't care about other fields it chooses not to care about. This does mean that another entity within the cluster can modify things without Strimzi's knowledge, but it shouldn't matter if the `metadata.managedFields` are chosen correctly.
+
+For users:
+
+* If you want to configure something for a single object (e.g. a single Kafka cluster), and it's in scope of the configuration, use that
+* If you want to configure something that's not the currently configurable fields, you can use other mechanisms to do so (Kyverno, other controllers), and Strimzi won't interfere
+* If you must configure something through some other method, and it's not directly being set in a different way by Strimzi, then Strimzi won't interfere and it'll "just work" (e.g. cattle annotations, pvc resizing annotations, metallb annotations. The user didn't make a choice about this other than the "I want to run Strimzi on this cluster that uses those technologies")
+
+This does not add the possibility of conflicts, they already exist. This reduces the liklihood of reaching a conflict and has some mechanisms in place to avoid them once they are found, although it is still possible if 2 controllers are forcing ownership on the same field of the same resource.
+
+### Three-way diff
+
+The live object (A) is stored in the `.metadata`, `.spec` etc. The knowledge of which fields the client previously applied, and indeed all clients (B) is stored in the `.metadata.managedFields`. And the incoming object (C) is supplied by the client. Kubernetes knows which fields were previously set by the client (in the case that the new object no longer specifies some, and they can be removed), and it knows about which client set all the other fields, and will keep them around instead of wiping out all the other labels etc.
```

Comment:
I still don't follow this part. How will we set which fields are managed by Strimzi. Will it be only the fields set to some value? Or will it be all fields which Strimzi allows users to configure even if they are null / not set?

I assume it will be the first one - but I think it should be clarified.

----

I think this also defines the backward incompatible change / Modifications to existing behaviours which is important for existing users to weigh in. E.g. does not setting for example `affinity` in the Kafka CR mean Strimzi enforces it to be null (i.e. it owns all fields it supports as configuration and enforces them to be not set if not set) or just ignores it (i.e. it owns only the values which are set and the values not set can be set by someone else without Strimzi overwriting them). 

I assume you want to ignore the unset fields since that was where we started this proposal. But this is an important change which I think has to be clearly described in the previous section. Right now, you cover there how will it work and what will be possible (which is ok and can stay there), but you do not touch on what actually changes for the users which I think is very important for understanding the consequences.